arangod/Aql/RemoteExecutor.cpp

```diff
@@ -442,46 +446,64 @@ Result ExecutionBlockImpl<RemoteExecutor>::sendAsyncRequest(fuerte::RestVerb typ
   network::ConnectionPool::Ref ref = pool->leaseConnection(spec.endpoint);
 
   std::lock_guard<std::mutex> guard(_communicationMutex);
-  unsigned ticket = ++_lastTicket;
+  auto ticket = generateNewTicket();
   std::shared_ptr<fuerte::Connection> conn = ref.connection();
   conn->sendRequest(std::move(req),
                     [=, ref(std::move(ref))](fuerte::Error err,
                                              std::unique_ptr<fuerte::Request>,
                                              std::unique_ptr<fuerte::Response> res) {
-                      _query.sharedState()->execute([&] {  // notifies outside
-                        std::lock_guard<std::mutex> guard(_communicationMutex);
-                        if (_lastTicket == ticket) {
-                          if (err != fuerte::Error::NoError || res->statusCode() >= 400) {
-                            _lastError = handleErrorResponse(spec, err, res.get());
-                          } else {
-                            _lastResponse = std::move(res);
-                          }
+                      std::lock_guard<std::mutex> guard(_communicationMutex);
```

Comment:
actually I do not get the need for the extra mutex here, an atomic _lastTicket should be sufficient so synchronize access to _lastError and _lastResponse ?

Reply:
The problem is that we do not only have to synchronize a request with its answer, but also protect an answer with another request; e.g., RemoteExecutor sends a getSome/skipSome request, then a shutdown happens (e.g. due to a timeout) at the same time when the answer arrives. Then, this code in `ExecutionBlockImpl<RemoteExecutor>::shutdown`
```
  if (!_hasTriggeredShutdown) {
    std::lock_guard<std::mutex> guard(_communicationMutex);
    std::ignore = generateNewTicket();
    _hasTriggeredShutdown = true;
  }
```
which resets `_lastTicket`, `_lastError` and `_lastResponse`, races with the lambda that's called with the answer:
```
                    [=, ref(std::move(ref))](fuerte::Error err,
                                             std::unique_ptr<fuerte::Request>,
                                             std::unique_ptr<fuerte::Response> res) {
                      std::lock_guard<std::mutex> guard(_communicationMutex);

                      if (_lastTicket == ticket) {
                        if (err != fuerte::Error::NoError || res->statusCode() >= 400) {
                          _lastError = handleErrorResponse(spec, err, res.get());
                        } else {
                          _lastResponse = std::move(res);
                        }
                        _query.sharedState()->execute();
                      }
                    }
```
which reads `_lastTicket` and then sets `_lastError` or `_lastResponse`. I do not see how that should be correctly synchronized with just an atomic `_lastTicket`.